### PR TITLE
add `-rejectparasites` option

### PR DESCRIPTION
**Pull request by Retropex**
can be considered a part of #64

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -648,6 +648,7 @@ void SetupServerArgs(ArgsManager& argsman)
                    OptionsCategory::NODE_RELAY);
     argsman.AddArg("-permitbaremultisig", strprintf("Relay non-P2SH multisig (default: %u)", DEFAULT_PERMIT_BAREMULTISIG), ArgsManager::ALLOW_ANY,
                    OptionsCategory::NODE_RELAY);
+    argsman.AddArg("-rejectparasites", strprintf("relay and mine parasites (default: %u)", DEFAULT_REJECT_PARASITES), ArgsManager::ALLOW_ANY, OptionsCategory::NODE_RELAY);
```
Comment (luke-jr) [inline src/init.cpp]: The description is backward

File: src/policy/policy.cpp [hunk 2]
```diff
@@ -177,6 +177,10 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
         if (IsDust(txout, dust_relay_fee)) {
             MaybeReject("dust");
         }
+
+        if (tx.nLockTime == 21 && reject_parasites == true){
+            MaybeReject ("parasite tx");
```
Comment (luke-jr) [inline src/policy/policy.cpp]: Both spaces should go (maybe just "parasite"?)

File: src/qt/optionsmodel.cpp [hunk 3]
```diff
@@ -1089,6 +1091,16 @@ bool OptionsModel::setOption(OptionID option, const QVariant& value, const std::
         }
         break;
     }
+    case rejectparasites:
+    {
+        if (changed()) {
+            const bool fNewValue = value.toBool();
+            node().mempool().m_reject_parasites = fNewValue;
+            // This option is inverted in the config:
```
Comment (luke-jr) [inline src/qt/optionsmodel.cpp]: It's not...

File: src/qt/optionsmodel.cpp [hunk 4]
```diff
@@ -1089,6 +1091,16 @@ bool OptionsModel::setOption(OptionID option, const QVariant& value, const std::
...
+            gArgs.ModifyRWConfigFile("rejectparasites", strprintf("%d", ! fNewValue));
```
Comment (luke-jr) [inline src/qt/optionsmodel.cpp]: Should only use the newer settings.json for new options

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/qt/optionsmodel.cpp (hunk 4 above)
Comment (Retropex) [reply src/qt/optionsmodel.cpp]: hum, I don't how to do that.

Comment (luke-jr) [reply src/qt/optionsmodel.cpp]: Check out rejectbarepubkey below

File: src/init.cpp [hunk 5]
```diff
@@ -648,6 +648,7 @@ void SetupServerArgs(ArgsManager& argsman)
                    OptionsCategory::NODE_RELAY);
     argsman.AddArg("-permitbaremultisig", strprintf("Relay non-P2SH multisig (default: %u)", DEFAULT_PERMIT_BAREMULTISIG), ArgsManager::ALLOW_ANY,
                    OptionsCategory::NODE_RELAY);
+    argsman.AddArg("-rejectparasites", strprintf("Parasites we relay and mine (default: %u)", DEFAULT_REJECT_PARASITES), ArgsManager::ALLOW_ANY, OptionsCategory::NODE_RELAY);
```
Comment (luke-jr) [inline src/init.cpp]:
```suggestion
    argsman.AddArg("-rejectparasites", strprintf("Refuse to relay or mine parasitic overlay protocols (default: %u)", DEFAULT_REJECT_PARASITES), ArgsManager::ALLOW_ANY, OptionsCategory::NODE_RELAY);
```

File: src/policy/policy.cpp [hunk 6]
```diff
@@ -177,6 +177,10 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
         if (IsDust(txout, dust_relay_fee)) {
             MaybeReject("dust");
         }
+
+        if (tx.nLockTime == 21 && reject_parasites == true){
+            MaybeReject("parasite");
```
Comment (luke-jr) [inline src/policy/policy.cpp]: Hmm, maybe even

```suggestion
            MaybeReject("parasite-cat21");
```

File: src/qt/optionsdialog.cpp [hunk 7]
```diff
@@ -263,6 +263,12 @@ OptionsDialog::OptionsDialog(QWidget* parent, bool enableWallet)
     verticalLayout_Spamfiltering->addWidget(rejectunknownscripts);
     FixTabOrder(rejectunknownscripts);
 
+    rejectparasites = new QCheckBox(groupBox_Spamfiltering);
+    rejectparasites->setText(tr("Reject parasites txs"));
+    rejectparasites->setToolTip(tr("With this option enabled, transactions related to parasites txs will be ignored. Parasites txs are transactions using Bitcoin as a technical infrastructure to animate other protocols, unrelated to the initially planned money transfers."));
```
Comment (luke-jr) [inline src/qt/optionsdialog.cpp]:
```suggestion
    rejectparasites->setToolTip(tr("With this option enabled, transactions related to parasitic overlay protocols will be ignored. Parasites are transactions using Bitcoin as a technical infrastructure to animate other protocols, unrelated to ordinary money transfers."));
```

File: src/qt/optionsmodel.cpp [hunk 8]
```diff
@@ -1089,6 +1091,15 @@ bool OptionsModel::setOption(OptionID option, const QVariant& value, const std::
         }
         break;
     }
+    case rejectparasites:
+    {
+        if (changed()) {
+            const bool fNewValue = value.toBool();
+            node().mempool().m_reject_parasites = fNewValue;
+            gArgs.ModifyRWConfigFile("rejectparasites", strprintf("%d", ! fNewValue));
```
Comment (luke-jr) [inline src/qt/optionsmodel.cpp]: This is still inverted too

File: src/qt/optionsmodel.h [hunk 9]
```diff
@@ -87,6 +87,7 @@ class OptionsModel : public QAbstractListModel
         incrementalrelayfee,
         mempoolexpiry,
         rejectunknownscripts,   // bool
+        rejectparasites,      // bool
```
Comment (luke-jr) [inline src/qt/optionsmodel.h]: nit: extra spaces to align the comments

File: src/test/fuzz/transaction.cpp [hunk 10]
```diff
@@ -70,8 +70,8 @@ FUZZ_TARGET(transaction, .init = initialize_transaction)
 
     const CFeeRate dust_relay_fee{DUST_RELAY_TX_FEE};
     std::string reason;
-    const bool is_standard_with_permit_bare_multisig = IsStandardTx(tx, std::nullopt, /* permit_bare_pubkey= */ true, /* permit_bare_multisig= */ true, dust_relay_fee, reason);
-    const bool is_standard_without_permit_bare_multisig = IsStandardTx(tx, std::nullopt, /* permit_bare_pubkey= */ true, /* permit_bare_multisig= */ false, dust_relay_fee, reason);
+    const bool is_standard_with_permit_bare_multisig = IsStandardTx(tx, std::nullopt, /* permit_bare_pubkey= */ true, /* permit_bare_multisig= */ true, /*reject_parasites*/ false, dust_relay_fee, reason);
```
Comment (luke-jr) [inline src/test/fuzz/transaction.cpp]: Include the `=` so Clang checks it

File: src/policy/policy.cpp [hunk 11]
```diff
@@ -177,6 +177,10 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
         if (IsDust(txout, dust_relay_fee)) {
             MaybeReject("dust");
         }
+
+        if (tx.nLockTime == 21 && reject_parasites == true){
```
Comment (luke-jr) [inline src/policy/policy.cpp]: This shouldn't be inside a `for` loop

File: src/validation.cpp [hunk 12]
```diff
@@ -767,7 +767,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
 
     // Rather not work on nonstandard transactions (unless -testnet/-regtest)
     std::string reason;
-    if (m_pool.m_require_standard && !IsStandardTx(tx, m_pool.m_max_datacarrier_bytes, m_pool.m_permit_bare_pubkey, m_pool.m_permit_bare_multisig, m_pool.m_dust_relay_feerate, reason, ignore_rejects)) {
```
Comment (luke-jr) [inline src/validation.cpp:770]: Why is `m_permit_bare_multisig` changing ` m_reject_parasites` here?

**Comment by luke-jr** [comment]
Please split the GUI changes out to a second commit

File: src/validation.cpp (hunk 12 above)
Comment (Retropex) [reply src/validation.cpp:770]: Probably a error with "search and replace"

**Comment by Retropex** [comment]
> Please split the GUI changes out to a second commit

done 👍

File: src/qt/optionsmodel.cpp [hunk 13]
```diff
@@ -1089,6 +1091,16 @@ bool OptionsModel::setOption(OptionID option, const QVariant& value, const std::
         }
         break;
     }
+    case rejectparasites:
+    {
+        if (changed()) {
+            // The config and internal option is inverted
+            const bool nv = ! value.toBool();
```
Comment (luke-jr) [inline src/qt/optionsmodel.cpp]: Incorrect

File: src/init.cpp [hunk 14]
```diff
@@ -648,6 +648,7 @@ void SetupServerArgs(ArgsManager& argsman)
                    OptionsCategory::NODE_RELAY);
     argsman.AddArg("-permitbaremultisig", strprintf("Relay non-P2SH multisig (default: %u)", DEFAULT_PERMIT_BAREMULTISIG), ArgsManager::ALLOW_ANY,
                    OptionsCategory::NODE_RELAY);
+    argsman.AddArg("-rejectparasites", strprintf("Refuse to relay or mine parasitic overlay protocols (default: %u)", DEFAULT_REJECT_PARASITES), ArgsManager::ALLOW_ANY, OptionsCategory::NODE_RELAY);
```
Comment (luke-jr) [inline src/init.cpp]: Let's move this down a few to be in somewhat alphabetical order?

File: src/policy/policy.cpp [hunk 15]
```diff
@@ -179,6 +179,10 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
         }
     }
 
+    if (tx.nLockTime == 21 && reject_parasites == true){
```
Comment (luke-jr) [inline src/policy/policy.cpp]: Missing space before `{`

Maybe check it before the loop for a quicker reject?

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/qt/optionsdialog.cpp [hunk 16]
```diff
@@ -263,6 +263,12 @@ OptionsDialog::OptionsDialog(QWidget* parent, bool enableWallet)
     verticalLayout_Spamfiltering->addWidget(rejectunknownscripts);
     FixTabOrder(rejectunknownscripts);
 
+    rejectparasites = new QCheckBox(groupBox_Spamfiltering);
+    rejectparasites->setText(tr("Reject parasites txs"));
```
Comment (luke-jr) [inline src/qt/optionsdialog.cpp]:
```suggestion
    rejectparasites->setText(tr("Reject parasite transactions"));
```

File: src/policy/policy.cpp (hunk 15 above)
Comment (Retropex) [reply src/policy/policy.cpp]: good idea

File: src/init.cpp (hunk 14 above)
Comment (Retropex) [reply src/init.cpp]: I lowered it from a line but we will also have to go up `-minrelaytxfee` for the order to be respected. (but it's out of scope of this PR)

File: src/policy/policy.h [hunk 17]
```diff
@@ -53,6 +53,8 @@ static constexpr unsigned int DEFAULT_WEIGHT_PER_DATA_BYTE{4};
 static constexpr bool DEFAULT_PERMIT_BAREPUBKEY{false};
 /** Default for -permitbaremultisig */
 static constexpr bool DEFAULT_PERMIT_BAREMULTISIG{false};
+/** Default for -rejectparasites*/
```
Comment (luke-jr) [inline src/policy/policy.h]: Missing space on the end

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/policy/policy.cpp [hunk 18]
```diff
@@ -149,6 +149,10 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
         }
     }
 
+    if (tx.nLockTime == 21 && reject_parasites == true) {
+        MaybeReject("parasite-cat21");
+    }
+
```
Comment (luke-jr) [inline src/policy/policy.cpp]: Would move this up yet further before the inputs loop

File: src/policy/policy.cpp [hunk 19]
```diff
@@ -149,6 +149,10 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
         }
     }
 
+    if (tx.nLockTime == 21 && reject_parasites == true) {
```
Comment (luke-jr) [inline src/policy/policy.cpp]:
```suggestion
    if (tx.nLockTime == 21 && reject_parasites) {
```

File: src/policy/policy.h [hunk 20]
```diff
@@ -53,6 +53,8 @@ static constexpr unsigned int DEFAULT_WEIGHT_PER_DATA_BYTE{4};
 static constexpr bool DEFAULT_PERMIT_BAREPUBKEY{false};
 /** Default for -permitbaremultisig */
 static constexpr bool DEFAULT_PERMIT_BAREMULTISIG{false};
+/** Default for -rejectparasites */
+static constexpr bool DEFAULT_REJECT_PARASITES{true};
```
Comment (luke-jr) [inline src/policy/policy.h:57]: Make this `false` for the first commit and add a third with it where `corepolicy` is also updated (eg https://github.com/bitcoinknots/bitcoin/commit/58c1e5ce547d61760a41e0febf68690c8e55d4bb)

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/init.cpp [hunk 21]
```diff
@@ -744,6 +744,7 @@ void InitParameterInteraction(ArgsManager& args)
         args.SoftSetArg("-bytespersigopstrict", "0");
         args.SoftSetArg("-permitbarepubkey", "1");
         args.SoftSetArg("-permitbaremultisig", "1");
+        args.SoftSetArg("-rejectparasites", "0");
```
Comment (luke-jr) [inline src/init.cpp:747]: The change to the DEFAULT_REJECT_PARASITES should be part of this commit too

Comment (Retropex) [reply src/init.cpp:747]: but I can't put the same change on two commit?

Comment (luke-jr) [reply src/init.cpp:747]: Add it as `false`, then change to `true` in this commit

**Comment by Retropex** [comment]
My bad, don't mind this force push, I will change it.

**Comment by Retropex** [comment]
should be good

**Comment by Retropex** [comment]
I have added tests lets me know if it's enough

**Comment by Retropex** [comment]
Merged in this commit: https://github.com/bitcoinknots/bitcoin/commit/1c1fcdba83ade749b2b301310c4a851553828572

**Pull request closed without merging**